### PR TITLE
Specify, review and approve operation Proposal-4

### PR DESCRIPTION
**Pull request by pavel-esir**
In the Proposal-4 added one additional output with probabilities of proposals. Also corrected a couple of typos in the previous specification. Proposal operation with 2 outputs is already enabled as 'extention' we want to add it to opset4.

JIRA: CVS-32964

### CODE REVIEW

File: docs/ops/detection/Proposal_4.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+*   **3**: 1D input floating tensor 3 or 4 elements:  [`image_height`, `image_width`, `scale_height_and_width`] or [`image_height`, `image_width`, `scale_height`, `scale_width`]. Required.
+
+**Outputs**:
+
+*   **1**: Floating point tensor of shape `[batch_size * post_nms_topn, 5]`.
+
+*   **2**: 1D floating point tensor of size `post_nms_topn` with probabilities. *Optional*
+
+**Example**
```
Comment (lazarevevgeny) [inline docs/ops/detection/Proposal_4.md:151]: Please, add a section called "Types". Refer to https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/sort/NonMaxSuppression_4.md for an example.

File: docs/ops/detection/Proposal_4.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+**Versioned name**: *Proposal-4*
+
+**Category**: *Object detection*
+
+**Short description**: *Proposal* operation filters bounding boxes and outputs only those with the highest prediction confidence.
+
+**Detailed description**
+
+*Proposal* has three inputs: a tensor with probabilities whether particular bounding box corresponds to background and foreground, a tensor with logits for each of the bounding boxes, a tensor with input image size in the [`image_height`, `image_width`, `scale_height_and_width`] or [`image_height`, `image_width`, `scale_height`, `scale_width`] format. The produced tensor has two dimensions `[batch_size * post_nms_topn, 5]`.
```
Comment (lazarevevgeny) [inline docs/ops/detection/Proposal_4.md:11]: Here the last statement is that the operation produces one tensor. It should be updated

File: docs/ops/detection/Proposal_4.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+*   **2**: 4D input floating point tensor with box logits. Required.
+
+*   **3**: 1D input floating tensor 3 or 4 elements:  [`image_height`, `image_width`, `scale_height_and_width`] or [`image_height`, `image_width`, `scale_height`, `scale_width`]. Required.
+
+**Outputs**:
+
+*   **1**: Floating point tensor of shape `[batch_size * post_nms_topn, 5]`.
+
+*   **2**: 1D floating point tensor of size `post_nms_topn` with probabilities. *Optional*
```
Comment (lazarevevgeny) [inline docs/ops/detection/Proposal_4.md:149]: "tensor of size `post_nms_topn` with probabilities" -> "tensor of shape `[post_nms_topn]` with probabilities"

File: docs/ops/detection/Proposal_4.md (hunk 1 above)
Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md:155]: Done.

File: docs/ops/detection/Proposal_4.md (hunk 2 above)
Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md]: Agree. In the last version updated this sentence.

File: docs/ops/detection/Proposal_4.md (hunk 3 above)
Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md]: Updated

**Comment by andrejsokolov** [comment]
@elatkin, please take a look at the PR.

File: docs/ops/detection/Proposal_4.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+* *T*: floating point type.
+
+**Example**
+
+```xml
+<layer ... type="Proposal" ... >
+    <data base_size="16" feat_stride="16" min_size="16" nms_thresh="0.6" post_nms_topn="200" pre_nms_topn="6000"
+     ratio="2.67" scale="4.0,6.0,9.0,16.0,24.0,32.0"/>
+    <input> ... </input>
+    <output> ... </output>
```
Comment (rkazants) [inline docs/ops/detection/Proposal_4.md:162]: please provide an example without reductions.

Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md]: Done. Please review

File: docs/ops/detection/Proposal_4.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+1.  Generates initial anchor boxes. Left top corner of all boxes is at (0, 0). Width and height of boxes are calculated from *base_size* with *scale* and *ratio* attributes.
+2.  For each point in the first input tensor:
+    *   pins anchor boxes to the image according to the second input tensor that contains four deltas for each box: for *x* and *y* of center, for *width* and for *height*
+    *   finds out score in the first input tensor
+3.  Filters out boxes with size less than *min_size*
+4.  Sorts all proposals (*box*, *score*) by score from highest to lowest
+5.  Takes top *pre_nms_topn* proposals
+6.  Calculates intersections for boxes and filter out all boxes with \f$intersection/union > nms\_thresh\f$
+7.  Takes top *post_nms_topn* proposals
+8.  Returns top proposals and optionally their probabilities
```
Comment (rkazants) [inline docs/ops/detection/Proposal_4.md:22]: please check that this part is actually rendered as a list

File: docs/ops/detection/Proposal_4.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    * *tensorflow* - calculate box coordinates like in the TensorFlow* Object Detection API models
+  * **Type**: string
+  * **Default value**: "" (empty string)
+  * **Required**: *no*
+
+**Inputs**:
+
+*   **1**: tensor of type *T* with class prediction scores. Required.
+
+*   **2**: tensor of type *T* with box logits. Required.
```
Comment (rkazants) [inline docs/ops/detection/Proposal_4.md:141]: please provide shapes for these inputs

File: docs/ops/detection/Proposal_4.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+*   **3**: tensor of type *T* with 3 or 4 elements:  `[image_height`, `image_width`, `scale_height_and_width]` or `[image_height`, `image_width`, `scale_height`, `scale_width]`. Required.
```
Comment (rkazants) [inline docs/ops/detection/Proposal_4.md:143]: 1-D tensor of type *T* with 3 or 4 elements: ...

File: docs/ops/detection/Proposal_4.md (hunk 5 above)
Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md:22]: Checked, it renders as expected.

File: docs/ops/detection/Proposal_4.md (hunk 7 above)
Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md]: Agree and information about dimension of tensors.

File: docs/ops/detection/Proposal_4.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+*   **2**: 4D tensor of type *T* with deltas for each bounding box. Required.
+
+*   **3**: 1D tensor of type *T* with 3 or 4 elements:  `[image_height`, `image_width`, `scale_height_and_width]` or `[image_height`, `image_width`, `scale_height`, `scale_width]`. Required.
+
+**Outputs**:
+
+*   **1**: tensor of type *T* and shape `[batch_size * post_nms_topn, 5]`.
+
+*   **2**: tensor of type *T* and shape `[post_nms_topn]` with probabilities. *Optional*.
```
Comment (lazarevevgeny) [inline docs/ops/detection/Proposal_4.md:149]: Multiply with batch size.

File: docs/ops/detection/Proposal_4.md (hunk 6 above)
Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md]: Done

File: docs/ops/detection/Proposal_4.md (hunk 8 above)
Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md]: Done.

File: docs/ops/detection/Proposal_4.md [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+**Types**
+
+* *T*: floating point type.
+
+**Example**
+
+```xml
+<layer ... type="Proposal" ... >
+    <data base_size="16" feat_stride="8" for_deformable="0" min_size="16" nms_thresh="1.0" normalize="0" post_nms_topn="1000" pre_nms_topn="1000" ratio="1" scale="1,2"/>
```
Comment (lazarevevgeny) [inline docs/ops/detection/Proposal_4.md:166]: Parameter 'for_deformable="0"' should be removed.

File: docs/ops/detection/Proposal_4.md [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    <input>
+        <port id="0">
+            <dim>1</dim>
```
Comment (lazarevevgeny) [inline docs/ops/detection/Proposal_4.md:169]: let's change the example so the input has batch size more than 1.

File: docs/ops/detection/Proposal_1.md [hunk 11]
```diff
@@ -31,16 +31,14 @@
   * **Required**: *yes*
 
 * *pre_nms_topn*
-
-  * **Description**: *pre_nms_topn* is the number of bounding boxes before the NMS operation. For example, *pre_nms_topn* equal to 15 means that the minimum box size is 15.
+  * **Description**: *pre_nms_topn* is the number of bounding boxes before the NMS operation. For example, *pre_nms_topn* equal to 15 means to take top 15 boxes with the highest scores.
```
Comment (lazarevevgeny) [inline docs/ops/detection/Proposal_1.md:34]: Please, update 'logits' -> 'box deltas' in this file also.

File: docs/ops/detection/Proposal_4.md (hunk 9 above)
Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md:166]: Removed.

File: docs/ops/detection/Proposal_1.md (hunk 11 above)
Comment (pavel-esir) [reply docs/ops/detection/Proposal_1.md:34]: Changed.

File: docs/ops/detection/Proposal_4.md (hunk 10 above)
Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md]: Changed to 3.

File: docs/ops/detection/Proposal_4.md [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+            <dim>28</dim>
+            <dim>28</dim>
+        </port>
+        <port id="2">
+            <dim>3</dim>
+        </port>
+    </input>
+    <output>
+        <port id="3" precision="FP32">
+            <dim>1000</dim>
```
Comment (lazarevevgeny) [inline docs/ops/detection/Proposal_4.md:186]: Should be 3000. The same for line 190.

Comment (lazarevevgeny) [reply docs/ops/detection/Proposal_4.md:186]: Actually, let's set the batch size to be 7 to not confuse with the shape of third input.

Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md:186]: Changed `batch_size` to 7.
>Should be 3000. 
Why? I don't understand

Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md:186]: This snippet of IR is from SSH network and it's `post_nms_topn` is equal to 1000. I advocate to leave it as it is.

Comment (lazarevevgeny) [reply docs/ops/detection/Proposal_4.md:186]: According to the specification the output shape is `[batch_size * post_nms_topn, 5]`

Comment (pavel-esir) [reply docs/ops/detection/Proposal_4.md]: 🤦sorry for that. Updated.
